### PR TITLE
ERC20Template4  - Confidential EVM

### PR DESCRIPTION
**Pull request by alexcos20**
### ERC20TemplateSapphire (asset files object stored in contract)
    - follows same functions and principles as ERC20TemplateEnterprise, with the following additions
    - should be deployed only on Oasis Sapphire, and all transactions should be encrypted, otherwise security is compromised
    - on initialize, files object(asset URLs) is stored in the contract
    - owner can change files object anytime, calling setFilesObject
    - for every order, consumer and serviceId are added to a mapping
    - has a list of allowed/denied providers (using standard ERC721.balanceOf). Contracts addresses can be changed by owner
    - when a provider tries to fetch the files object (by calling getFilesObject), the following conditions are checked:
        - provider address has to be in allow list and not on deny list (given those lists are not address_zero)
        - consumer has to have a valid order

### CODE REVIEW

**Comment by trentmc** [comment]
I recommend renaming it from "ERC20TemplateSapphire.sol" --> "ERC20Template4.sol".

Here's why:

First, calling it "ERC20TemplateSapphire.sol" is a misnomer. Since template 3 is also Sapphire-only. And it's tuned for Predictoor, of course.

Second, remember the problem with the name "ERC20TemplateEnterprise.sol": it was a template useful beyond enterprises too. However the "enterprise" portion always made people think it was just for enterprises.

Third: there's a good chance that we deploy >=1 new variant of template 3 for Predictoor. Eg for continuous-valued predictions, or other new features into the contract.  Similarly, there's a good chance we deploy variants of the other templates too (including this new one). It starts to get messy to manage if the names are misleading. Hence why there's a simple name, plus a table to show what characteristics each template has.

If these weren't smart contracts, because of the issues above, we probably would have refactored the names into "ERC20Template1.sol", "ERC20Template2.sol", and "ERC20Template3.sol" long ago. But we only realized them when getting into template 3, and that's why it's already named ERC20Template3.sol". Let's keep this trend going: I recommend that this this new template, call it "ERC20Template4.sol".

**Review by trentmc — CHANGES_REQUESTED** [review]
See my two comments in the PR conversation: (a) update the table (b) change name to ERCTemplate4.sol.

**Comment by alexcos20** [comment]
> Hey @alexcos20 I recommend to also update the README that describes the templates: https://github.com/oceanprotocol/contracts/tree/main/contracts/templates.

Updated in https://github.com/oceanprotocol/contracts/pull/870/files#diff-e3d04a7f6d6d8ac50c0603b74dee727caa6d28993abdaeef48552bd29ac83dcc

**Comment by alexcos20** [comment]
@trentmc - renamed in https://github.com/oceanprotocol/contracts/pull/870/commits/80e682a3160282fae3bbf2ef6dd99fb6311d6565

File: contracts/templates/README.md [hunk 1]
```diff
@@ -40,11 +40,22 @@ Ocean-contracts v1.x (Ocean V4 “Onda”) is built on top of templates, in orde
...
+    - on initialize, files object(asset URLs) is stored in the contract
+    - owner can change files object anytime, calling setFilesObject
+    - for every order, consumer and serviceId are added to a mapping
+    - has a list of allowed/denied providers (using standard ERC721.balanceOf). Contracts addresses can be changed by owner
+    - when a provider tries to fetch the files object (by calling getFilesObject), the following conditions are checked:
+        - provider address has to be in allow list and not on deny list (given those lists are not address_zero)
+        - consumer has to have a valid order
 
 The following table sums it up
 Template # | Label | Allows fre/dispense by default? | Allows custody of datatokens? | Combines txs? | Purpose
```
Comment (trentmc) [inline contracts/templates/README.md:55]: The table should have >=1 columns that distinguish template 2 from template 4. 

Suggestion: "file object itself in contract". Where the value is "n" for template 1, 2, 3; and "y" for template 4.

**Review by trentmc — CHANGES_REQUESTED** [review]

File: contracts/templates/README.md (hunk 1 above)
Comment (alexcos20) [reply contracts/templates/README.md]: done

**Pull request merged**